### PR TITLE
Adding Auth0 authentication to Project page

### LINKED ISSUE #61: Determine an authentication method
Write a specification that describes how users will authenticate given the decision to use Auth0. 

- [x] AC: describes what identity(ies) users can use
- [x] AC: describes how users will authenticate in the browser
- [x] AC: describes how users will authenticate from the command line. 
- [ ] AC: Includes TOS (get from Shawn)

### LINKED ISSUE #62: Design a logout method
Design a log-out method or page on the data browser. (includes experience)

### LINKED ISSUE #150: Design a authentication method/experience

### LINKED ISSUE #153: Browser frontend  Auth0 authentication integration
# User story
Users need to be authenticated in the browser before they can view the project page(maybe more).

Relates to #155 

## Definition of done
- [ ] Add a login and signup button
- [ ] Use auth0 Authorization Code flow to authenticate users
- [ ] Restrict access to the project page when the user it not authenticated
- [ ] Redirect a user to login if they try to go to a page that requires authentication. Redirect a user back to their original page after login.
- [ ] Provide logout button.
- [ ] Store the access token.

# links
- https://auth0.com/blog/securing-gatsby-with-auth0/
- https://auth0.com/docs/libraries/auth0-spa-js/migrate-from-auth0js
- https://www.gatsbyjs.org/docs/node-apis/

### LINKED ISSUE #155: Setup Auth0
- [ ] Add browser as an application
- [ ] Customize Log in flow to show TOS and require user consent
- [ ] Setup google as an identity provider
- [ ] Setup Frontend to authenticate using auth0

### LINKED ISSUE #170: Determine the authentication-related user flows
- [x] Find out from Sean how much we can allow non-authenticated users to view:
- nothing
- list-of-projects page
- project page sans download file area
- all of project page including download area, but they must authenticate to get URLs

- [x] Document how the system and user will interact as a user flow

### PR DESCRIPTION
**Pull request by Bento007**
If the user is not authenticated and tries to visit a project page, they will be directed to login before they can proceed. 

This code is based off these tutorials:
- https://auth0.com/blog/securing-gatsby-with-auth0/?_ga=2.87703896.1785923316.1582573428-1657564689.1581035098
- https://www.gatsbyjs.org/tutorial/authentication-tutorial/
- https://auth0.com/blog/authenticating-your-first-react-app/

## Issue Links
- Closes #61, 
- Closes #62 
- Closes #150, 
- Closes #155,
- Closes #153, 
- Closes #170

## Instructions
1. run `gatsby develop`
1. open browser to localhost 
Screen Shot 2020-03-13 at 3 40 42 PM
1. view project 
Screen Shot 2020-03-13 at 3 48 04 PM
1. See download section
Screen Shot 2020-03-16 at 9 43 21 AM

1. select Login/Register 
Screen Shot 2020-03-13 at 3 41 17 PM
1. login 
Screen Shot 2020-03-13 at 3 41 59 PM

## I need help completing these tasks bellow
- DONE: Send the authorization token in the request for project info.
- DONE: Return to original page after login.
- DONE: Add a login/logout/register button

## Future Work
DONE: Use auth0-spa-js instead of auth0-js for authentication
Future TODO: Enforce authorization header in requests to backend.

### CODE REVIEW

File: dcp_prototype/frontend/src/components/login-signup.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import React from "react"
+import { useAuth0 } from "../contexts/auth0Context"
+import { Heading } from "theme-ui"
+
+const LoginSignup = () => {
+  const { isAuthenticated, loading, loginWithRedirect } = useAuth0()
```
Comment (calvinnhieu) [inline dcp_prototype/frontend/src/components/login-signup.js]: Wondering if there's a pattern we can use that initializes `auth0` when necessary (e.g. per page as opposed to per component). Just something to think about. OK with this long term if the overhead of calling `useAuth0()` is minimal.

File: dcp_prototype/frontend/gatsby-ssr.js [hunk 2]
```diff
@@ -4,4 +4,9 @@
  * See: https://www.gatsbyjs.org/docs/ssr-apis/
  */
 
-// You can delete this file if you're not using it
+import React from "react"
+import { Auth0Provider} from "./src/contexts/auth0Context"
```
Comment (calvinnhieu) [inline dcp_prototype/frontend/gatsby-ssr.js]: spacing

File: dcp_prototype/frontend/src/components/login-signup.js (hunk 1 above)
Comment (Bento007) [reply dcp_prototype/frontend/src/components/login-signup.js]: From my limit react and gatsby knowledge, it is initialized per page. Using the `wrapRootElement` in _gatsby-browser.js_ and _gatsby-node.js_ we wrap every page in an Auth0 context that initializes Auth0 for every page. The `loading` flag is used to determine if the async response from Auth0 has completed.

File: dcp_prototype/frontend/gatsby-ssr.js (hunk 2 above)
Comment (Bento007) [reply dcp_prototype/frontend/gatsby-ssr.js]: fixed

**Pull request merged**